crates/proto/src/rr/dnssec/proof.rs

```diff
@@ -0,0 +1,421 @@
+    #[error("ds response empty: {name}")]
+    DsResponseEmpty {
+        /// No records for the DS query were returned
+        name: Name,
+    },
+
+    /// DS record does not exist, and this was proven with an NSEC
+    #[error("ds record does not exist: {name}")]
+    DsResponseNsec {
+        /// The name of the DS record
+        name: Name,
+    },
+
+    /// The DnsKey is not marked as a zone key
+    #[error("not a zone signing key: {name} key_tag: {key_tag}")]
+    NotZoneDnsKey {
+        /// Name of the DNSKEY
+        name: Name,
+        /// The key tag derived from the DNSKEY
+        key_tag: u16,
+    },
+
+    /// There was a protocol error when looking up DNSSEC records
+    #[error("communication failure for query: {query}: {proto}")]
+    Proto {
+        /// Query that failed
+        query: Query,
+        /// Resons fo the failure
+        proto: ProtoError,
+    },
+
+    /// The RRSIGs for the rrset are not present.
+    ///    It's indeterminate if DS records can't be found
+    ///    It's bogus if the DS records are present
+    #[error("rrsigs are not present for: {name} record_type: {record_type}")]
+    RrsigsNotPresent {
+        /// Name that RRSIGS are missing for
+        name: Name,
+        /// The record type in question
+        record_type: RecordType,
+    },
+
+    /// The RRSIGs could not be verified or failed validation
+    #[error("rrsigs were not able to be verified: {name}, type: {record_type}")]
+    RrsigsUnverified {
+        /// Name that RRSIGS failed for
+        name: Name,
+        /// The record type in question
+        record_type: RecordType,
+    },
+
+    /// The self-signed dnskey is invalid
+    #[error("self-signed dnskey is invalid: {name}")]
+    SelfSignedKeyInvalid {
+        /// Name of the DNSKEY
+        name: Name,
+    },
+}
+
+/// The error type for dnssec errors that get returned in the crate
+#[derive(Debug, Clone, Error)]
+pub struct ProofError {
+    /// The proof derived from the failed state
+    pub proof: Proof,
+    /// The kind of error
+    pub kind: ProofErrorKind,
+}
+
+impl ProofError {
+    /// Create an error with the given Proof and Associated Error
+    pub fn new(proof: Proof, kind: ProofErrorKind) -> Self {
+        Self { proof, kind }
+    }
+
+    /// Get the kind of the error
+    pub fn kind(&self) -> &ProofErrorKind {
+        &self.kind
+    }
+}
+
+impl fmt::Display for ProofError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}: {}", self.proof, self.kind)
+    }
+}
+
+/// A wrapper type to ensure that the state of a DNSSEC proof is evaluated before use
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Proven<T> {
+    proof: Proof,
+    value: T,
+}
+
+impl<T> Proven<T> {
+    /// Wrap the value with the given proof
+    pub fn new(proof: Proof, value: T) -> Self {
+        Self { proof, value }
+    }
+
+    /// Get the associated proof
+    pub fn proof(&self) -> Proof {
```

Comment:
it would, but it does weaken the `Proven` type, which I do like. For example, you can't get at the interior of `Option` or `Result` without being forced to consider the status, and I like the parallel to that here.

My needs are really around legacy/interop on the way to eventually using only hickory, and I'd hate to be the guy that compromised it from the start! :-)

Reply:
So, decompose would still end up with a type that could be unused. (The Result above requires evaluation).

A different option could be to make the Proof type itself `#[must_use]`? Then we could do the `decompose` and there would be a warning if `Proof` is not evaluated?

Reply:
And I think at that point it's basically just an `unwrap` like method or `parts` could be another...

Reply:
I think `decompose()` would more commonly be called `into_parts()`.

Reply:
ok, I added `Proven::into_parts` that should close out this discussion.